ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -0,0 +1,400 @@
+//! The Yk JIT IR
+
+use std::fmt;
+use strum_macros::FromRepr;
+
+/// Number of bits used to encode an opcode.
+const OPCODE_SIZE: u64 = 8;
+
+/// Max number of operands in a short instruction.
+const SHORT_INSTR_MAX_OPERANDS: u64 = 3;
+
+/// Bit fiddling for a short operands.
+const SHORT_OPERAND_SIZE: u64 = 18;
+const SHORT_OPERAND_KIND_SIZE: u64 = 3;
+const SHORT_OPERAND_KIND_MASK: u64 = 7;
+const SHORT_OPERAND_MASK: u64 = 0x3ffff;
+
+/// Bit fiddling for instructions.
+const INSTR_ISSHORT_SIZE: u64 = 1;
+const INSTR_ISSHORT_MASK: u64 = 1;
+const INSTR_OPCODE_MASK: u64 = 0xe;
+
+/// An instruction is identified by its index in the instruction vector.
+pub(crate) type InstructionID = usize;
+
+/// An operand kind.
+#[repr(u64)]
+#[derive(Debug, FromRepr, PartialEq)]
+pub enum OpKind {
+    /// The operand is not present.
+    ///
+    /// This is used in short instructions where 3 operands are inlined. If the instruction
+    /// requires fewer then 3 operands, then it can use this variant to express that.
+    ///
+    /// By using the zero discriminant, this means that a freshly created short instruction has
+    /// with zero operands until they are explicitly filled in.
+    NotPresent = 0,
+    /// The operand references a previously defined local variable.
+    Local,
+}
+
+impl From<u64> for OpKind {
+    fn from(v: u64) -> Self {
+        // unwrap safe assuming only valid discriminant numbers are used.
+        Self::from_repr(v).unwrap()
+    }
+}
+
+#[derive(Debug, FromRepr, PartialEq)]
+#[repr(u64)]
+pub enum OpCode {
+    Load,
+    LoadArg,
+}
+
+impl From<u64> for OpCode {
+    fn from(v: u64) -> Self {
+        // unwrap safe assuming only valid discriminant numbers are used.
+        Self::from_repr(v).unwrap()
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Operand {
+    Long(LongOperand),
+    Short(ShortOperand),
+}
+
+impl Operand {
+    pub(crate) fn new(kind: OpKind, val: u64) -> Self {
+        if val < 2u64.pow(15) {
```

Comment:
Hang on, isn't this exceeding the range of `u64`? What is the aim here? I wonder if what was intended was to reserve `u64::MAX` as a special value?

Reply:
It's checking if the operand's payload can fit in a small operand, where we have only 15-bits of wiggle room.

(a long instruction, once implemented, would allow larger operands)

I missed the 15 in my "kill all magic numbers" pass too.

Reply:
(2^15 should fit in a u64 though, or I've gone mad)

Reply:
Right, got it. I'll leave the comment here so we remember to resolve it.

Reply:
>  (2^15 should fit in a u64 though, or I've gone mad)

I misread it because I'm so unfamiliar with that way of doing bit patterns -- mea culpa!

Reply:
Added a constant in a34b500956cfb5a71d3a63fd2f4eb78e582ce7b7